src/drivers/uavcan_v1/Uavcan.cpp

```diff
@@ -0,0 +1,310 @@
+
+	if (_instance == nullptr) {
+		PX4_ERR("Out of memory");
+		return -1;
+	}
+
+	const int node_init_res = 0; //_instance->init(node_id, can->driver.updateEvent());
+
+	if (node_init_res < 0) {
+		delete _instance;
+		_instance = nullptr;
+		PX4_ERR("Node init failed %i", node_init_res);
+		return node_init_res;
+	}
+
+	// Keep the bit rate for reboots on BenginFirmware updates
+	_instance->active_bitrate = bitrate;
+
+	_instance->ScheduleOnInterval(ScheduleIntervalMs * 1000);
+
+	return PX4_OK;
+}
+
+void UavcanNode::Run()
+{
+	pthread_mutex_lock(&_node_mutex);
+
+	if (!_initialized) {
+		// interface init
+		if (_can_interface) {
+			if (_can_interface->init() == PX4_OK) {
+
+				// Subscribe to messages uavcan.node.Heartbeat.
+				canardRxSubscribe(&_canard_instance,
+						  CanardTransferKindMessage,
+						  uavcan::node::Heartbeat_1_0::PORT_ID,  // The fixed Subject-ID of the Heartbeat message type (see DSDL definition).
+						  uavcan::node::Heartbeat_1_0::SIZE,      // The maximum payload size (max DSDL object size) from the DSDL definition.
+						  CANARD_DEFAULT_TRANSFER_ID_TIMEOUT_USEC,
+						  &_heartbeat_subscription);
+
+
+				// canardRxSubscribe(&_canard_instance,   // Subscribe to messages uavcan.node.Heartbeat.
+				// 		  CanardTransferKindMessage,
+				// 		  REGULATED_DRONE_SENSOR_BMSSTATUS_ID,  // The fixed Subject-ID of the Heartbeat message type (see DSDL definition).
+				// 		  uavcan::node::Heartbeat_1_0::SIZE,      // The maximum payload size (max DSDL object size) from the DSDL definition.
+				// 		  CANARD_DEFAULT_TRANSFER_ID_TIMEOUT_USEC,
+				// 		  &_drone_sensor_BMSStatus_subscription);
+
+				_initialized = true;
+			}
+		}
+
+		// return early if still not initialized
+		if (!_initialized) {
+			pthread_mutex_unlock(&_node_mutex);
+			return;
+		}
+	}
+
+	// check for parameter updates
+	if (_parameter_update_sub.updated()) {
+		// clear update
+		parameter_update_s pupdate;
+		_parameter_update_sub.copy(&pupdate);
+
+		// update parameters from storage
+	}
+
+	perf_begin(_cycle_perf);
+	perf_count(_interval_perf);
+
+	CanardFrame received_frame;
+
+	if (_can_interface->receive(&received_frame) > 0) {
+
+		CanardTransfer receive;
+		int32_t result = canardRxAccept(&_canard_instance, &received_frame, 0, &receive);
+
+		if (result < 0) {
+			// An error has occurred: either an argument is invalid or we've ran out of memory.
+			// It is possible to statically prove that an out-of-memory will never occur for a given application if
+			// the heap is sized correctly; for background, refer to the Robson's Proof and the documentation for O1Heap.
+			// Reception of an invalid frame is NOT an error.
+			PX4_ERR("Receive error %d\n", result);
+
+		} else if (result == 1) {
+			// A transfer has been received, process it.
+			_canard_instance.memory_free(&_canard_instance, (void *)receive.payload); // Deallocate the dynamic memory afterwards.
+
+			if (receive.port_id == REGULATED_DRONE_SENSOR_BMSSTATUS_ID) {
```

Comment:
It's still in flux, but I've since updated the BMSStatus case to get the port id from a PX4 parameter. In fixes cases like Heartbeat I've added it as part of the generated code.